lib/galaxy/model/store/__init__.py

```diff
@@ -0,0 +1,1321 @@
+                new_obj = obj.copy()
+                new_obj["id"] = new_id
+                return (k, new_obj)
+
+            return (k, obj)
+
+        params = job_attrs['params']
+        params = remap(params, remap_objects)
+        return params
+
+
+class BagArchiveImportModelStore(DirectoryImportModelStoreLatest):
+
+    def __init__(self, bag_archive, **kwd):
+        archive_dir = tempfile.mkdtemp()
+        bdb.extract_bag(bag_archive, output_path=archive_dir)
+        # Why this line though...?
+        archive_dir = os.path.join(archive_dir, os.listdir(archive_dir)[0])
+        bdb.revert_bag(archive_dir)
+        super(BagArchiveImportModelStore, self).__init__(archive_dir, **kwd)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ModelExportStore(object):
+
+    @abc.abstractmethod
+    def export_history(self, history, include_hidden=False, include_deleted=False):
+        """Export history to store."""
+
+    @abc.abstractmethod
+    def add_dataset_collection(self, collection):
+        """Add HDCA to export store."""
+
+    @abc.abstractmethod
+    def add_dataset(self, dataset, include_files=True):
+        """Add HDA to export store."""
+
+    @abc.abstractmethod
+    def __enter__(self):
+        """Export store should be used as context manager."""
+
+    @abc.abstractmethod
+    def __exit__(self):
+        """Export store should be used as context manager."""
+
+
+class DirectoryModelExportStore(ModelExportStore):
+
+    def __init__(self, export_directory, app=None, for_edit=False, serialize_dataset_objects=None, export_files=None):
+        if not os.path.exists(export_directory):
+            os.makedirs(export_directory)
+
+        if app is not None:
+            self.app = app
+            self.security = app.security
+        else:
+            self.security = IdEncodingHelper(id_secret="randomdoesntmatter")
+
+        self.export_directory = export_directory
+        self.serialization_options = model.SerializationOptions(
+            for_edit=for_edit,
+            serialize_dataset_objects=serialize_dataset_objects,
+            serialize_files_handler=self,
+        )
+        self.export_files = export_files
+        self.included_datasets = {}
+        self.included_collections = []
+        self.included_libraries = []
+        self.included_library_folders = []
+        self.collection_datasets = {}
+        self.collections_attrs = []
+        self.dataset_id_to_path = {}
+
+    def serialize_files(self, dataset, as_dict):
+        if self.export_files is None:
+            return None
+        elif self.export_files == "symlink":
+            add = os.symlink
+        elif self.export_files == "copy":
+            def add(src, dest):
+                if os.path.isdir(src):
+                    shutil.copytree(src, dest)
+                else:
+                    shutil.copyfile(src, dest)
+
+        export_directory = self.export_directory
+        if dataset.id in self.included_datasets:
```

Comment:
So it would seem, good catch. I've eliminated the conditional in https://github.com/galaxyproject/galaxy/pull/7684 (which already had a bunch of model store fixes/enhancements).